### PR TITLE
perf(kernel): optimize gfx950 gluon MHA extend kernel

### LINKED ISSUE #521: Tokenspeed kernel amd integration
### Checklist

- [x] I searched existing issues before filing this report.
- [x] I am running on a supported GPU — NVIDIA: H100, H200, B200, B300, GB200, GB300; or AMD: MI350X, MI355X. (Other hardware is not supported and issues will be closed.)
- [ ] NVIDIA users: I am using CUDA 13 and CUDA driver ≥ 580 (recommended). If not, we may not be able to reproduce the issue.

### Summary

The decode kernel does not seem to be able to handle physical page size id > 65536
```
  from tokenspeed_kernel_amd.ops.attention.gluon.mha_prefill_fp16_gfx950 import (
      gluon_mha_decode_fp16_gfx950,
  )
```

### Reproduction

I am using `vllm/vllm-openai-rocm:nightly` docker image and 

```bash
git clone https://github.com/lightseekorg/tokenspeed
cd tokenspeed/tokenspeed-kernel-amd/python
python3 -m pip install -e .
```

The following code hangs `python repro_decode_large_page_id.py --page-id 65536`

`repro_decode_large_page_id.py`

```python
from __future__ import annotations

import argparse
import sys

import torch

from tokenspeed_kernel_amd.ops.attention.gluon.mha_decode_fp16_gfx950 import (
    gluon_mha_decode_fp16_gfx950,
)


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--page-id",
        type=int,
        default=65536,
        help="Physical KV page id to put in page_table.",
    )
    parser.add_argument("--page-size", type=int, default=64)
    parser.add_argument("--num-q-heads", type=int, default=64)
    parser.add_argument("--num-kv-heads", type=int, default=8)
    parser.add_argument("--head-dim", type=int, default=64)
    parser.add_argument(
        "--dtype",
        choices=("float16", "bfloat16"),
        default="bfloat16",
    )
    parser.add_argument(
        "--device",
        default="cuda",
        help="Use CUDA_VISIBLE_DEVICES to select the physical ROCm GPU.",
    )
    parser.add_argument(
        "--sliding-window",
        type=int,
        default=-1,
        help="Use -1 for full attention; non-negative for sliding decode.",
    )
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    if not torch.cuda.is_available():
        raise RuntimeError("This repro requires a ROCm/CUDA-visible GPU.")

    dtype = torch.bfloat16 if args.dtype == "bfloat16" else torch.float16
    device = torch.device(args.device)
    num_blocks = args.page_id + 1

    elements_per_page = args.page_size * args.num_kv_heads * args.head_dim
    first_element_offset = args.page_id * elements_per_page
    print(f"device={device} name={torch.cuda.get_device_name(device)}", flush=True)
    print(f"dtype={dtype}", flush=True)
    print(f"num_blocks={num_blocks}", flush=True)
    print(f"page_id={args.page_id}", flush=True)
    print(f"elements_per_page={elements_per_page}", flush=True)
    print(f"first_element_offset={first_element_offset}", flush=True)
    print(f"int32_max={2**31 - 1}", flush=True)
    print(
        "expected="
        + ("FAIL/overflow-risk" if first_element_offset >= 2**31 else "PASS")
        ,
        flush=True,
    )

    print("allocating tensors", flush=True)
    q = torch.zeros(
        (1, args.num_q_heads, args.head_dim), device=device, dtype=dtype
    )
    k_cache = torch.zeros(
        (num_blocks, args.page_size, args.num_kv_heads, args.head_dim),
        device=device,
        dtype=dtype,
    )
    v_cache = torch.zeros_like(k_cache)

    # Make the selected page nonzero so a successful run has a visible result.
    print("initializing selected page", flush=True)
    k_cache[args.page_id].fill_(0.01)
    v_cache[args.page_id].fill_(1.0)

    page_table = torch.tensor([[args.page_id]], device=device, dtype=torch.int32)
    cache_seqlens = torch.tensor([args.page_size], device=device, dtype=torch.int32)

    print("launching decode", flush=True)
    torch.cuda.synchronize()
    out = gluon_mha_decode_fp16_gfx950(
        q=q,
        k_cache=k_cache,
        v_cache=v_cache,
        page_table=page_table,
        cache_seqlens=cache_seqlens,
        max_seqlen_k=args.page_size,
        max_seqlen_q=1,
        window_left=args.sliding_window,
        sinks=None,
    )
    torch.cuda.synchronize()

    print(f"output_shape={tuple(out.shape)}", flush=True)
    print(f"output_sum={out.float().sum().item():.6f}", flush=True)
    print("completed", flush=True)
    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except Exception as exc:
        print(f"FAILED: {type(exc).__name__}: {exc}", file=sys.stderr)
        raise
```

### Environment

I am integrating the kernel into vLLM at https://github.com/vllm-project/vllm/pull/46742

### PR DESCRIPTION
**Pull request by Yu-Zhewen**
Fix https://github.com/lightseekorg/tokenspeed/issues/521#issuecomment-4986417504.

- Packed GQA layout — pack the query-head group into the MFMA M dim (BLOCK_M = BLOCK_Q × GROUP_SIZE) so each KV tile is loaded once per group instead of once per query head, killing the GROUP_SIZE× KV reload.
- Ragged token-based grid — grid over true q-block count with in-kernel cu_seqlens_q self-location, so a q=1 request costs exactly one block (no rectangular over-provisioning) while staying CUDA-graph static.
- Split-K (full attention only) — for low-occupancy long-KV cases, split each request's KV stream across programs (compute pass writes per-split partials + LSE) and merge in a reduce pass.

## Extend kernel: latency on the 6 mixed workloads (ms, gfx950 / MI355X)

| workload                    | before (main) | now   | AITER | now/AITER (<1 faster) |
|-----------------------------|--------------:|------:|------:|----------------------:|
| extend_full_q1              | 0.199         | 0.044 | 0.064 | 0.69                  |
| extend_full_q8              | 0.194         | 0.049 | 0.126 | 0.39                  |
| extend_sliding128_q8        | 0.050         | 0.023 | 0.030 | 0.78                  |
| mixed_balanced              | 0.224         | 0.065 | 0.088 | 0.74                  |
| mixed_balanced_sliding128   | 0.044         | 0.023 | 0.027 | 0.87                  |
| mixed_prefill_heavy         | 2.896         | 0.315 | 0.352 | 0.89                  |

*64 q-heads, 8 KV-heads, head_dim 64, bf16, sinks on; median of 100 iters.*

Signed-off-by: Yu-Zhewen <zhewenyu@amd.com>

### CODE REVIEW

File: tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/attention/gluon/mha_extend_gfx950.py [hunk 1]
```diff
@@ -56,12 +43,54 @@
...
+    pid_qblock,
+    pid_batch,
+):
+    # Resolve (batch, q_pos_base) from the launch grid:
+    #   * rectangular ``(blocks_per_req, batch, n_kv_heads)``: ``pid_qblock`` is
+    #     the block within the request and ``pid_batch`` the request.
+    #   * ragged ``(total_num_q_blocks, ...)``: ``pid_qblock`` is a global q-block
+    #     that binary-searches ``cu_seqlens_q`` to find its request.
+    if RAGGED:
+        batch = _find_seq_idx(cu_seqlens_q_ptr, pid_qblock, num_seqs, BLOCK_Q)
```
Comment (borontion) [inline tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/attention/gluon/mha_extend_gfx950.py:84]: can you explain a little bit about what's the purpose of this `_find_seq_idx`?

Comment (Yu-Zhewen) [reply tokenspeed-kernel-amd/python/tokenspeed_kernel_amd/ops/attention/gluon/mha_extend_gfx950.py:84]: Previously, we had the rectangular grid `(blocks_per_req, batch, n_kv_heads)`. Every request gets the same `blocks_per_req = ceil(max_seqlen_q / BLOCK_Q`) blocks, which is inefficient in a mixed batch (e.g. many q=1 decode-like requests alongside a few long prefills). Instead, here we flatten the first two dimension to get a ragged grid `(total_num_q_blocks, n_kv_heads)`.  `program_id(0)` is now a flat block index over the whole batch. The program no longer knows its request from the grid coordinates, and we use `_find_seq_idx` to recover it.
